app/controllers/v0/sign_in_controller.rb

```diff
@@ -229,13 +229,13 @@ def validate_authorize_params(type, client_id, acr, operation)
       if client_config(client_id).blank?
         raise SignIn::Errors::MalformedParamsError.new message: 'Client id is not valid'
       end
-      unless SignIn::Constants::Auth::CSP_TYPES.include?(type)
+      unless client_config(client_id).valid_credential_service_provider?(type)
```

Comment:
❓ Does this memoization pattern work? I didn't think you could memoize a method that accepted an argument without jumping through some hoops. I **think** the code as written would look up the client config multiple times, but I could be wrong.

```ruby
    def client_config(client_id)
      @client_config ||= SignIn::ClientConfig.find_by(client_id:)
    end
```

Reply:
I ran a test in IRB to check, and it looks like it does memoize the lookup:

```
irb(main):001* def client_config(client_id)
irb(main):002*   @client_config ||= client_id
irb(main):003> end
=> :client_config
irb(main):004> client_config(1)
=> 1
irb(main):005> @client_config
=> 1
irb(main):006> client_config(2)
=> 1
irb(main):007> @client_config
=> 1
```

I guess the part that is throwing me off is that we pass in the `client_id` for each call, even though the value of `@client_config` won't change regardless of what we pass in.

Maybe it would be more idiomatic to do something like this:

```ruby
config = client_config(client_id)
if config.blank?
  raise SignIn::Errors::MalformedParamsError.new message: 'Client id is not valid'
end
unless config.valid_credential_service_provider?(type)
  raise SignIn::Errors::MalformedParamsError.new message: 'Type is not valid'
end
unless config.valid_service_level?(acr)
  raise SignIn::Errors::MalformedParamsError.new message: 'ACR is not valid'
end
```

Reply:
I wonder why the client config is memoized at all. It eems like something we could initialize with a `before_action`.  Any ideas on why we originally went the memoization route @bosawt?

Reply:
Client config used to be a set of definitions in `Settings` and then was moved to a database-based solution, so my best guess is a result of that transition?

Reply:
I think we could discuss refactoring the memoization but maybe in another PR. Do you want to make a ticket and propose a new way to do it? There are more functions than the `authorize` function that uses this, so would be good to resolve it for the entire controller